### PR TITLE
Jetpack section: add rules to control the appearance of the server credentials banner

### PR DESCRIPTION
**Pull request by rcanepa**
#### Changes proposed in this Pull Request

* Control the appearance of the banner according to a very specific set of rules. 
* The banner will never show up if any of these conditions becomes true:
  * The banner was dismissed (clicking the CTA is considered dismissing the banner).
  * The banner has appeared at least ten times.
  * The first banner appearance happened more than 90 days ago.

#### Questions

* The ticket mentions that we should only show the banner for existing customers (meaning customers that have seen the existing credentials form in Settings > Security). What about users that never saw the form in Settings > Security but they went there because they saw a guide/tutorial online? I'm sure many new users learn about WordPress from seeing YouTube videos and reading articles.

#### Testing instructions

Note: this test suite is separated in three parts, one for each condition. Never refresh the page until you have completed the three parts.

Prerequisite: a self-host Jetpack site with a paid plan or the Scan product.

##### Condition: The banner has appeared at least ten times.
* Run this PR and enable the `jetpackSidebarSection` AB test.
* Go to `http://calypso.localhost:3000` and select your Jetpack site.
* Go to Settings > Security.
* Go to Settings > Jetpack and back to Settings > Security 10 times. We want to increase the `view` count by ten.
* Verify that you no longer see the banner.

##### Condition: The banner was dismissed (clicking the CTA is considered dismissing the banner).
* Open the Console and execute `state.preferences.remoteValues['backup-scan-security-settings-moved'] = []` to reset the count to 0.
* Go to Settings > Jetpack and then to Settings > Security (this will save in the back-end the reset we just did).
* In the Console, execute `state.preferences.remoteValues['backup-scan-security-settings-moved']` and verify that the result has only one object of type `view`. 
* Click the banner CTA.
* Go back to Settings > Security.
* Verify that you no longer see the banner.
* In the Console, execute `state.preferences.remoteValues['backup-scan-security-settings-moved']` and verify that the result has an object of type `dismiss`. 

##### Condition: The first banner appearance happened more than 90 days ago.
* Open the Console and execute `state.preferences.remoteValues['backup-scan-security-settings-moved'] = []` to reset the count to 0.
* Go to Settings > Jetpack and then to Settings > Security without refreshing the page (this will save in the back-end the reset we just did).
* In the Console, execute `state.preferences.remoteValues['backup-scan-security-settings-moved']` and verify that the result has only one object of type `view`.
* In the Console, execute `state.preferences.remoteValues['backup-scan-security-settings-moved'][0].date = new Date('2006-12-21T04:30:11+00:00').getTime()`. This will change the date of the first view to the year 2006 (feel free to play with this date and change it to something else).
* Go to Settings > Jetpack and then back to Settings > Security.
* Verify that you no longer see the banner.
* In the Console, execute `state.preferences.remoteValues['backup-scan-security-settings-moved']` and verify that the result has no object of type `dismiss` and the objects of type `view` are less than 10.

Fixes 1179060693083348-as-1180806262315505.

##### Demo
Kapture 2020-06-23 at 23 58 50

### CODE REVIEW

**Comment by keoshi** [comment]
> Condition: The banner has appeared at least ten times.

I'm getting a `We're sorry, but an unexpected error has occurred` error and this output in the console:

image

**Comment by rcanepa** [comment]
> > Condition: The banner has appeared at least ten times.
> 
> I'm getting a `We're sorry, but an unexpected error has occurred` error and this output in the console, when accessing `Settings > Security`:
> 
> image

Thanks for catching this. I was initializing a variable as `{}` when it should have been an `[]`. It should be good for another review now :)

**Comment by keoshi** [comment]
One note on the testing script:

> Condition: The first banner appearance happened more than 90 days ago.
> - Go to Settings > Jetpack and then to Settings > Jetpack

I believe this should be:

- Go to Settings > Jetpack and then to Settings > Security

**Comment by rcanepa** [comment]
> One note on the testing script:
> 
> > Condition: The first banner appearance happened more than 90 days ago.
> > 
> > * Go to Settings > Jetpack and then to Settings > Jetpack
> 
> I believe this should be:
> 
> * Go to Settings > Jetpack and then to Settings > Security

Instruction updated. Thanks! 🙇

File: client/my-sites/site-settings/jetpack-credentials-banner/index.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+	return (
+		<>
+			<button
+				onClick={ () => dispatch( savePreference( JETPACK_CREDENTIALS_BANNER_PREFERENCE, {} ) ) }
+			>
+				Reset{ ' ' }
+			</button>
+			<Banner
+				callToAction="Take me there!"
```
Comment (ChaosExAnima) [inline client/my-sites/site-settings/jetpack-credentials-banner/index.tsx:61]: Should this be translated?

File: client/state/site-settings/jetpack-credentials-banner/selectors.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	const currentDateInMilli = Date.now();
+	const firstViewInMilli = firstView.date;
+	if ( firstViewInMilli + NINETY_DAYS_IN_MILLISECONDS >= currentDateInMilli ) {
+		return false;
+	}
+
+	return true;
+}
+
+const MAX_BANNER_VIEWS = 10;
```
Comment (ChaosExAnima) [inline client/state/site-settings/jetpack-credentials-banner/selectors.ts:73]: May be cleaner to have all constants at the top of the file?

File: client/state/site-settings/jetpack-credentials-banner/selectors.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+ * Returns if the banner should be displayed to the user. The results depends on three
+ * conditions:
+ * 1. The view count is less than 10.
+ * 2. The banner has not been dismissed by the user.
+ * 3. The first time the user saw the banner was less than 90 days ago.
+ *
+ * @param  {object}  state  Global state tree
+ * @returns {boolean}       Whether the banner should be displayed
+ */
+export function shouldDisplayJetpackCredentialsBanner( state ): boolean {
```
Comment (ChaosExAnima) [inline client/state/site-settings/jetpack-credentials-banner/selectors.ts:84]: Should `state` use a Redux type?

File: client/my-sites/site-settings/jetpack-credentials-banner/types.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+export type PreferenceType = 'dismiss' | 'view';
+
+export type Preference = {
+	date: number;
+	type: PreferenceType;
+};
```
Comment (ChaosExAnima) [inline client/my-sites/site-settings/jetpack-credentials-banner/types.ts:6]: Love this! Though shouldn't this be [`types.d.ts`](https://microsoft.github.io/TypeScript-New-Handbook/chapters/type-declarations/)? Or does that not play nice with our webpack config?

File: client/my-sites/site-settings/jetpack-credentials-banner/index.tsx (hunk 1 above)
Comment (rcanepa) [reply client/my-sites/site-settings/jetpack-credentials-banner/index.tsx:61]: Yes. Thanks!

File: client/my-sites/site-settings/jetpack-credentials-banner/types.ts (hunk 4 above)
Comment (rcanepa) [reply client/my-sites/site-settings/jetpack-credentials-banner/types.ts:6]: Not sure, I'll try with `types.d.ts`.

**Comment by rcanepa** [comment]
>I tried to test but getting not found errors for state in the console. Am I missing anything?

Oh, probably there is no access to `state` in Horizon.

>Otherwise, code looks good aside from a few things I flagged.

Nice! I'll work on those and ping you back.

File: client/my-sites/site-settings/jetpack-credentials-banner/types.ts (hunk 4 above)
Comment (rcanepa) [reply client/my-sites/site-settings/jetpack-credentials-banner/types.ts]: This worked. Now the file is called `types.d.ts`.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/3926115

Thank you @rcanepa for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/3926115).

**Pull request merged**